### PR TITLE
Improve `ZSTM`'s performance

### LINKED ISSUE #8983: Using STM is much slower than using a single Ref even with contention
Here's the result of `SingleRefBenchmark`:
```
[info] Benchmark                     (fibers)  (ops)   Mode  Cnt     Score     Error  Units
[info] STMBenchmarks.refContention         10   1000  thrpt    5  4854.440 ± 300.593  ops/s
[info] STMBenchmarks.trefContention        10   1000  thrpt    5   440.238 ±   5.423  ops/s
```
Also if I add a single ref case to `TMapContentionBenchmarks`:
```
[info] Benchmark                      Mode  Cnt    Score   Error  Units
[info] STMBenchmarks.contentionRef   thrpt    5  204.426 ± 2.632  ops/s
[info] STMBenchmarks.contentionTMap  thrpt    5   29.248 ± 1.729  ops/s
[info] STMBenchmarks.contentionTRef  thrpt    5   33.547 ± 1.804  ops/s
```
In both cases, using a single `Ref` is an order of magnitude faster than STM even in a contention case, which makes STM loses its appeal.

### PR DESCRIPTION
**Pull request by kyri-petrou**
/fixes #8983
/claim #8983

/cc @jdegoes

This PR is a fairly major rewrite of ZSTM's transactional runtime, which improves its performance so that the STM-based classes such as `TRef` and `TMap` to have a performance that is comparable with the non-STM ones.

> [!NOTE]
> If you never heard of ZSTM before or you're unsure about how it can be used in an application, make sure to check out [this fantastic blog post](https://blog.pierre-ricadat.com/the-tri-z-architecture-a-pattern-for-layering-zio-applications-in-scala) by @ghostdogpr!

## Issue

The main issue with the previous implementation of `ZSTM.atomically` (i.e., `ZSTM#commit`) was that it relied on running code within `synchronized` blocks on a global object. This resulted in really bad performance especially under contention, because we wouldn't only be synchronizing against the current ZSTM instance, but across all of them that would be running at the same time within the JVM (even if totally unrelated!).

The reason that the global locks were needed is that ZSTM data types are essentially just `TRef`s combined in different ways. When we commit a change, we need to ensure that for all of the ZSTMs we read/write to/from, their state wasn't modified after we read its original value. Since it's not possible to know in advance which `TRef`s a ZSTM is using, the global lock was used to ensure that no thread would modify the state of any `TRef` (even if unrelated to our current transaction!!) by the use of synchronization.

## Solution

The solution I adopted in this PR was for each TRef to contain its own `ReentrantLock`. Besides `ReentrantLock`s being more lightweight than `synchronized` blocks, they also have the following advantages:
1. They offer _optional_ locking via `tryLock`, so we can simply retry the transaction if we can't acquire the lock, as it's _very_ likely that some other thread is already attempting to commit a transaction.
2. They are virtual-thread friendly; in fact, users that use the Loom executor will very likely see an even bigger benefit out of this change

When we run a `ZSTM` and therefore know which `TRef`s we are reading from / writing to, we can then acquire a lock on each of them before proceeding to commit the changes. Simple, right?

Well, it's not _this_ simple. Remember that we said we can't know which `TRef`s our ZSTM is using until we run it? Well, depending on the state of the ZSTM when we read it, we can end up using different `TRef`s. e.g., when we need to resize the underlying `TArray` in a `TMap`, we are modifying more `TRef`s than when we are not. So the exact same ZSTM that writes a value in a `TMap` might use different number `TRef`s 🫠. To account for this, if we locked on `TRef`s that were determined in a previous `run` invocation, we need to check that we still hold the lock on all of them after we run our ZSTM.

Now, the second issue with this approach is that if 2 threads try to acquire locks on `TRefA` and `TRefB`, and each thread locks one of them, we might end up in a deadlock. To account for this, we:
1. ~~`ReentrantLock#tryLock` with a timeout instead of `ReentrantLock#lock`. This way when one of the threads times out, the other thread can then acquire the lock and continue.~~
2. ~~Use a random timeout value to reduce contention if multiple threads try to acquire the lock at the exact same time (PS: I'm considering to remove this)~~
3. Use `ReentrantLock(fair = true)`, so that threads that have been waiting the longest to acquire the lock are prioritised. This improves performance especially for `TMap` (which uses at least 2 `TRef`s) since the likelihood of a deadlock happening waiting on a timeout is much less

**EDIT**: Thanks to @mschuwalow's recommendation, it seems we can avoid the need for (1) and (2) above if we sort the locks before acquiring them. For this, I ended up using a `TreeMap` as the underlying implementation of `Journal`. This simplifies the code massively, but also improves performance substantially (~20% throughput improvement under heavy contention).

Besides the above, I also optimized a bunch of CPU hotpaths and the ZSTM "runloop" based on optimizations that I previously did to [improve the performance of ZPure](https://github.com/zio/zio-prelude/pull/1306) (in zio-prelude) that turned out to work great in both benchmarks and real applications (as confirmed by @ghostdogpr)

## Benchmark results

Ok I blabbered for long enough, here's a few screenshots for the `SingleRefBenchmark` benchmark:

Up to ~800% throughput performance under heavy contention including multiple unrelated `TRef`s compared to series/2.x
image

A fairly substantial reduction in allocations (I'm actually not very certain what I optimized to get this 😅)
image

And comparing with `TRef` with `Ref`, we get better performance under contention (note that Ref still performs better when there's no contention but I guess that's expected)
image

### Full benchmark results

For the rest of the benchmarks / results, head over to https://jmh.morethan.io/ and upload these 2 JSONS 👇 

[results-main.json](https://github.com/user-attachments/files/16511583/results-main.json)
[results-pr.json](https://github.com/user-attachments/files/16511587/results-pr.json)

Let me know what you think!

### CODE REVIEW

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -1539,11 +1539,14 @@ sealed trait ZIO[-R, +E, +A]
...
-      def loop(n: Int): ZIO[R, E, A] =
-        self.flatMap(a => if (n <= 0) Exit.succeed(a) else ZIO.yieldNow *> loop(n - 1))
-
-      loop(n)
+      if (n0 <= 0) self
+      if (n0 == 1) {
+        self *> self
+      } else {
+        ZIO.whileLoop(n0 != 0)(self)(_ => n0 -= 1) *> self
+      }
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/ZIO.scala]: While I was running the benchmarks, I realised the performance of `ZIO#repeatN` was quite bad. This improves it massively. What do we think of it?

File: core/shared/src/main/scala-2.12/zio/stm/ZSTMUtils.scala [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package zio.stm
+
+import scala.collection.mutable
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala-2.12/zio/stm/ZSTMUtils.scala]: in your experience, are scala mutable collections on par with java ones performance-wise?

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 1 above)
Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/ZIO.scala]: 👍

Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/ZIO.scala]: PS: Opened #9082 to track this separately

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 3]
```diff
@@ -761,92 +761,107 @@ sealed trait ZSTM[-R, +E, +A] extends Serializable { self =>
...
+            case effect: Effect[Any, Any, Any] =>
+              val a = effect.f(journal, fiberId, env)
+              if (contStack eq Nil) exit = TExit.Succeed(a, onCommitStack)
+              else {
+                curr = contStack.head(a)
+                contStack = contStack.tail
+              }
+
+            case onSuccess: OnSuccess[Any, Any, Any, Any] =>
+              contStack ::= onSuccess.k
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala:803]: what is that `::=`? 🤔

File: core/jvm/src/main/scala-3/zio/stm/ZSTMLockSupport.scala [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.concurrent.{ThreadLocalRandom, TimeUnit}
+
+private object ZSTMLockSupport {
+
+  type Lock = ReentrantLock
+  object Lock {
+    def apply(fair: Boolean = false): Lock = new ReentrantLock(fair)
+  }
+
+  inline def lock[A](refs: collection.Set[TRef[?]])(inline f: A)(implicit rnd: ThreadLocalRandom): Boolean =
```
Comment (hearnadam) [inline core/jvm/src/main/scala-3/zio/stm/ZSTMLockSupport.scala]: As a general question, should we use `using/given` in Scala 3 specific files?

File: core/shared/src/main/scala/zio/stm/TRef.scala [hunk 5]
```diff
@@ -35,10 +36,20 @@ import java.util.concurrent.atomic.AtomicReference
...
+  @volatile private[stm] var todo: Map[TxnId, Todo] = Map.empty
 ) extends Serializable { self =>
 
+  @deprecated("kept for binary compatibility only", "2.1.8")
+  def this(versioned: Versioned[A], todo: AtomicReference[Map[TxnId, Todo]]) =
+    this(new AtomicReference(versioned.value), todo.get())
+
+  private[stm] val lock: ZSTMLockSupport.Lock = ZSTMLockSupport.Lock(true)
+
+  private[this] val hc = super.hashCode()
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/stm/TRef.scala:50]: can drop `[this]`?

File: core/shared/src/main/scala/zio/stm/TRef.scala [hunk 6]
```diff
@@ -142,18 +152,15 @@ final class TRef[A] private (
...
-    }
+    journal.getOrElseUpdate(self, newEntry())
 
   private[stm] def unsafeGet(journal: Journal): A =
     getOrMakeEntry(journal).unsafeGet
 
   private[stm] def unsafeSet(journal: Journal, a: A): Unit =
     getOrMakeEntry(journal).unsafeSet(a)
+
+  private[this] val newEntry = () => Entry(self.asInstanceOf[TRef[A & AnyRef]])
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/stm/TRef.scala:164]: Is there a reason to define methods as `val func = () =>` vs `def func()`

File: core/shared/src/main/scala-2.12/zio/stm/ZSTMUtils.scala [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+private object ZSTMUtils {
+
+  def newMutableMap[K, V](expectedNumElements: Int): mutable.HashMap[K, V] = {
+    val map = new mutable.HashMap[K, V]
+    if (expectedNumElements > 12) {
```
Comment (ghostdogpr) [inline core/shared/src/main/scala-2.12/zio/stm/ZSTMUtils.scala]: Maybe add a comment with the rationale for those numbers?

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 1 above)
Comment (ghostdogpr) [reply core/shared/src/main/scala/zio/ZIO.scala]: Sounds good but I would separate it from this PR.

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 8]
```diff
@@ -1563,13 +1593,13 @@ object ZSTM {
       b.flatMap(b => if (b) stm.asSome else none)
   }
 
-  private[stm] final case class FailException[E](e: E) extends Throwable(null, null, false, false)
+  private[stm] final case class FailException[E](e: E) extends ControlThrowable
```
Comment (ghostdogpr) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala:1596]: 🙏🙏🙏

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 9]
```diff
@@ -1617,10 +1647,13 @@ object ZSTM {
   private[zio] def succeedNow[A](a: A): USTM[A] = SucceedNow(a)
 
   private[stm] object internal {
-    val DefaultJournalSize = 4
-    val MaxRetries         = 10
-    val YieldOpCount       = 2048
+    final val DefaultJournalSize   = 3
```
Comment (ghostdogpr) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala:1651]: Curious why 4 -> 3?

Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1651]: Previously we used a java.util.HashMap and passed `4` in the constructor. Because the constructors of Scala's HashMap are different for 2.12 and 2.13+, I had to use a version-specific method that uses the intended number of elements than the actual capacity. So 3 / 0.75d = 4

File: core/jvm/src/main/scala-2/zio/stm/ZSTMLockSupport.scala [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+  @inline def tryLock[A](lock: Lock)(f: => A): Boolean = {
+    val lock0 = lock
+    if (lock0.tryLock()) {
+      try { f; true }
+      finally lock0.unlock()
+    } else false
+  }
+
+  @inline private def tryLockN[A](refs: collection.Set[TRef[?]], size: Int)(f: => A): Unit = {
```
Comment (regiskuckaertz) [inline core/jvm/src/main/scala-2/zio/stm/ZSTMLockSupport.scala]: You don't need to produce the result `locked` here?

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 11]
```diff
@@ -1632,282 +1665,265 @@ object ZSTM {
...
+        else if (value.isChanged) JournalAnalysis.ReadWrite
+        else JournalAnalysis.ReadOnly
+
+      private[this] def analyzeJournalN(): JournalAnalysis = {
+        var changed = false
+
+        val it = map.valuesIterator
+        while (it.hasNext) {
+          val value = it.next()
+          if (value.isInvalid) return JournalAnalysis.Invalid
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala:1727]: haven't seen that in a while 😄 non-local returns are deprecated in scala 3, it uses exceptions under the hood so the performance is not optimal. better use an `isValid` var instead.

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 12]
```diff
@@ -1632,282 +1665,265 @@ object ZSTM {
...
-          val oldTodo = todo.get
+      /**
+       * Flag indicating whether the journal is valid
+       *
+       * '''NOTE''': This method MUST be invoked while we hold the lock on the
+       * journal
+       */
+      private[ZSTM] def isValid: Boolean = {
+        val it = map.valuesIterator
+        while (it.hasNext) if (!it.next().isValid) return false
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala:1779]: Same as above

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 13]
```diff
@@ -1632,282 +1665,265 @@ object ZSTM {
...
+      }
 
-          loop = !todo.compareAndSet(oldTodo, emptyTodo)
+      /**
+       * Creates a function that can reset the journal.
+       */
+      private[ZSTM] def resetFn(): () => Unit = {
+        val currentNewValues = ZSTMUtils.newMutableMap[TRef[_], Any](map.size)
+        val itCapture        = map.iterator
+        while (itCapture.hasNext) {
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala:1789]: maybe this whole map initialization could be moved into the `newMutableMap` helper

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 14]
```diff
@@ -1632,282 +1665,265 @@ object ZSTM {
...
+        () => {
+          val saved = ZSTMUtils.newMutableMap[TRef[_], Entry](map.size)
+          val it    = map.iterator
+          while (it.hasNext) {
+            val (key, value) = it.next()
+            val resetValue = currentNewValues.getOrElse(key, null) match {
+              case null  => value.expected
+              case value => value
+            }
+            saved.update(key, value.copy().reset(resetValue))
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala:1803]: inline this into the pattern match above, it seems you coud avoid ony copy/reset in the `null` branch?

File: core/jvm/src/main/scala-2/zio/stm/ZSTMLockSupport.scala (hunk 10 above)
Comment (kyri-petrou) [reply core/jvm/src/main/scala-2/zio/stm/ZSTMLockSupport.scala]: We could, although all the usages of it just modify vars in the `f` and discard the boolean. Although I'll change this for consistency

File: core/shared/src/main/scala-2.12/zio/stm/ZSTMUtils.scala (hunk 7 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala-2.12/zio/stm/ZSTMUtils.scala]: Actually I did this because I thought Scala 2.12's HashMap initializes the map with a size of 16 like in Java / Scala 2.13+ but it doesn't seem create the underlying arrays at all. I'll change this to always size it

File: core/shared/src/main/scala/zio/stm/TRef.scala (hunk 6 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/stm/TRef.scala:164]: I found this out while looking at decompiled code.

The `getOrElseUpdate(key, newValue)` method on Scala's `HashMap`, has the second param as `=> A`. Since `=> A` is just syntactic sugar for `() => A`, this way we're not allocating or creating any closures when we call `getOrElseUpdate`.

File: core/shared/src/main/scala/zio/stm/TRef.scala (hunk 5 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/stm/TRef.scala:50]: The Scala 2 `private` and `private[this]` generate different bytecode. When `[this]` is not used, the compiler will create accessor methods for the val so that other instances of the class can access it. e.g., the code below:

```scala
class Hello {
  private val foo = "hello"
  private[this] val bar = "world"
}
```
<details>
  <summary>Produces the following bytecode (Click to expand)</summary>

```java
// class version 52.0 (52)
// access flags 0x21
public class Hello {

  // compiled from: Main.scala

  @Lscala/reflect/ScalaSignature;(bytes="\u0006\u0005\u00012A!\u0002\u0004\u0001\u0013!)\u0001\u0003\u0001C\u0001#!9A\u0003\u0001b\u0001\n\u0013)\u0002B\u0002\u0010\u0001A\u0003%a\u0003\u0003\u0004 \u0001\u0001\u0006IA\u0006\u0002\u0006\u0011\u0016dGn\u001c\u0006\u0002\u000f\u00059A(Z7qift4\u0001A\n\u0003\u0001)\u0001\"a\u0003\u0008\u000e\u00031Q\u0011!D\u0001\u0006g\u000e\u000cG.Y\u0005\u0003\u001f1\u0011a!\u00118z%\u00164\u0017A\u0002\u001fj]&$h\u0008F\u0001\u0013!\u0009\u0019\u0002!D\u0001\u0007\u0003\r1wn\\\u000b\u0002-A\u0011q\u0003H\u0007\u00021)\u0011\u0011DG\u0001\u0005Y\u0006twMC\u0001\u001c\u0003\u0011Q\u0017M^1\n\u0005uA\"AB*ue&tw-\u0001\u0003g_>\u0004\u0013a\u00012be\u0002")

  ATTRIBUTE ScalaSig : unknown

  ATTRIBUTE ScalaInlineInfo : unknown

  // access flags 0x12
  private final Ljava/lang/String; foo

  // access flags 0x12
  private final Ljava/lang/String; bar

  // access flags 0x2
  private foo()Ljava/lang/String;
   L0
    LINENUMBER 2 L0
    ALOAD 0
    GETFIELD Hello.foo : Ljava/lang/String;
    ARETURN
   L1
    LOCALVARIABLE this LHello; L0 L1 0
    MAXSTACK = 1
    MAXLOCALS = 1

  // access flags 0x1
  public <init>()V
   L0
    LINENUMBER 1 L0
    ALOAD 0
    INVOKESPECIAL java/lang/Object.<init> ()V
   L1
    LINENUMBER 2 L1
    ALOAD 0
    LDC "hello"
    PUTFIELD Hello.foo : Ljava/lang/String;
   L2
    LINENUMBER 3 L2
    ALOAD 0
    LDC "world"
    PUTFIELD Hello.bar : Ljava/lang/String;
   L3
    LINENUMBER 1 L3
    RETURN
   L4
    LOCALVARIABLE this LHello; L0 L4 0
    MAXSTACK = 2
    MAXLOCALS = 1
}
```
</details>

(see how there is a `private foo()` method but not `private bar()`)

File: core/shared/src/main/scala/zio/stm/ZSTM.scala (hunk 11 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1727]: This is not a non-local return :) Non-local returns are when the `return` keyword is contained within a Function. When they're used in a method directly they're just normal return statements.

I explained this in a bit more detail in this PR https://github.com/zio/zio/pull/9016#discussion_r1680690960 previously if you're interested

File: core/jvm/src/main/scala-3/zio/stm/ZSTMLockSupport.scala (hunk 4 above)
Comment (kyri-petrou) [reply core/jvm/src/main/scala-3/zio/stm/ZSTMLockSupport.scala]: I thought of it, the only issue with it is that when we use `using`, we can't pass the implicit like this raises a compiler error: 

```scala
lock(???)(???)(rnd)
```

It has to be changed to this to work which we can't cross-compile

```scala 3
lock(???)(???)(using rnd)
```

But since here we're not passing the implicit anywhere I think we can change it

File: core/shared/src/main/scala/zio/stm/ZSTM.scala (hunk 3 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:803]: Syntactic sugar for `contStack = onSuccess.k :: contStack` (contStack is List `var`)

File: core/shared/src/main/scala-2.12/zio/stm/ZSTMUtils.scala (hunk 2 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala-2.12/zio/stm/ZSTMUtils.scala]: It's extremely case-specific. Usually if you need to convert to/from Scala's immutable collections (which we do a lot in this file), Scala's mutable collections are faster.

Otherwise, I found them to be quite case-specific depending on the nature of the keys. I don't remember exactly under what conditions, but sometimes Java's is faster while other times Scala's is faster

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 1 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/ZIO.scala]: Sounds good, I'll revert this change for now

File: core/shared/src/main/scala/zio/stm/ZSTM.scala (hunk 13 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1789]: I don't think we can since the new map has a different `V` type than the original map

File: core/shared/src/main/scala/zio/stm/ZSTM.scala (hunk 14 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1803]: It seems that we _might_ be able to avoid the `copy()` but not the `reset` (tests are failing if I don't reset in the `null` case).

However, this method is only used for `ZSTM#orElse` so I'd like to play it safe here even if slightly less performant just in case there's an edge case that we're not capturing in the test suite. WDUT?

File: core/shared/src/main/scala/zio/stm/ZSTM.scala (hunk 9 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1651]: Added a comment with an explanation on why 3

**Comment by mschuwalow** [comment]
In regards to deadlocks due to
```
Now, the second issue with this approach is that if 2 threads try to acquire locks on TRefA and TRefB, and each thread locks one of them, we might end up in a deadlock. To account for this, we:
```

What about having a unique, orderable id per TRef (coming from some global counter)? When we lock the refs for the transaction, when can then acquire the locks in order of these numbers, ensuring that we always lock TRefs in the same order and avoiding these deadlocks.

**Comment by kyri-petrou** [comment]
> What about having a unique, orderable id per TRef (coming from some global counter)? When we lock the refs for the transaction, when can then acquire the locks in order of these numbers, ensuring that we always lock TRefs in the same order and avoiding these deadlocks.

Hmm this might work. I'm not sure if we can rely on this entirely or whether we'll still need the timeouts, but even if we do need the timeouts it'll definitely help with reducing contention. I'll give it a go!

**Comment by kyri-petrou** [comment]
@mschuwalow thanks a lot of the recommendation! I ended up using a `TreeMap` (SortedMap) to store the `TRef`s (and subsequently, to lock them) and not only that removed the need for lock timeouts to avoid the deadlock issue, but it also improved performance considerably under contention (~20%) 🎉

I'll update the PR description to match the current implementation

File: core/shared/src/main/scala/zio/stm/ZSTM.scala (hunk 11 above)
Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1727]: oh, interesting that the semantics differs based on the use site, thanks for the details!

File: core/shared/src/main/scala/zio/stm/ZSTM.scala (hunk 14 above)
Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1803]: sure, I think if the reset is necessary then the copy definitely is too

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 15]
```diff
@@ -16,14 +16,13 @@
 
 package zio.stm
 
-import zio._
-import zio.internal.{Stack, Sync}
 import zio.stacktracer.TracingImplicits.disableAutoTrace
+import zio.{FiberId, _}
```
Comment (ghostdogpr) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala]:
```suggestion
import zio._
```

**Review by ghostdogpr — DISMISSED** [review]
Fantastic work!

File: core/shared/src/main/scala/zio/stm/TRef.scala (hunk 5 above)
Comment (987Nabil) [reply core/shared/src/main/scala/zio/stm/TRef.scala:50]: @kyri-petrou is private[this] not deprecated in scala 3? 🤔

File: core/shared/src/main/scala/zio/stm/TRef.scala (hunk 6 above)
Comment (987Nabil) [reply core/shared/src/main/scala/zio/stm/TRef.scala:164]: I had a hunch that this is the case, but was not sure. Good to know 👍

File: core/shared/src/main/scala/zio/stm/TRef.scala (hunk 5 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/stm/TRef.scala:50]: > @kyri-petrou is private[this] not deprecated in scala 3? 🤔

@987Nabil It depends on what `-source` scalac flag you use. It raises a compiler warning if you're using `-source:future` but doesn't if you leave it unspecified

**Review by regiskuckaertz — APPROVED** [review]
💪

**Comment by jdegoes** [comment]
@kyri-petrou 

I love your work here and the performance improvements speak for themselves! The fact that it's Loom friendly -- and will not actually block OS threads under Loom -- is also a huge bonus for the Loom community.

I'm going to approve and merge this, however, I would also put up another bounty for trying a somewhat different approach -- basically having a single fiber dedicated to doing all updates, with all transactions executing the commit portion on that fiber. 

This world would involve no `ReentrantLock`, and everything would be async, Loom or not.

Let me know if there's any interest in exploring that.

**Comment by jdegoes** [comment]
Btw, sorting the locks is indeed the correct way to solve the deadlock issue. 👍 Thanks @mschuwalow for the input here!

**Comment by kyri-petrou** [comment]
@jdegoes Thanks for approving and merging it in!

> basically having a single fiber dedicated to doing all updates, with all transactions executing the commit portion on that fiber

This is something I explored a little bit as part of this PR, however the issue that I run into was that since we can't know in advance which `TRef`s a ZSTM is using until we run it, we ended up with a single thread doing all the work across all ZSTM instances. This ended up being more or less the same bottleneck as using a global lock in the old implementation, albeit without hard-blocking OS threads.

Having said that, I think this definitely has the potential of working really well in servers with low number of CPU available or when we have very few ZSTMs in our program. Or perhaps we need to re-write ZSTMs so that we can partition them without running them.
